### ISSUE TITLE
ACME Renewal Information (ARI) support

### ISSUE THREAD

**Issue by bt90** [caddyserver/caddy#5516]
LetsEncrypt added a new API to query renewal information:

https://letsencrypt.org/2023/03/23/improving-resliiency-and-reliability-with-ari.html

IETF draft for the ACME extension:

https://datatracker.ietf.org/doc/draft-ietf-acme-ari/

**Comment by francislavoie** [comment]
Yeah, it's been worked on https://github.com/mholt/acmez/commit/9f1389192a30cda988b52590124511ba248e2f38

During implementation though, it was found that ARI doesn't really make sense. There's been a bunch of discussion in the LE forums about it where Matt pointed out some problems.

**Comment by mholt** [comment]
My [first topic](https://community.letsencrypt.org/t/ari-how-do-you-get-the-der-encoded-certid-asn-1-sequence-from-x509-certificate-in-go/195353) in the LE forums was about how to produce the proper payload. Turns out it's almost the same thing as an OCSP request, but a little different, and the code already exists in our ocsp dependency, but it's not exported, so I had to repeat some logic (didn't really copy verbatim -- I still rewrote it -- but close).

The [second topic](https://community.letsencrypt.org/t/can-ari-conforming-clients-be-granted-exemptions-to-relevant-rate-limits/195600?u=mholt) I raised is more serious: ARI is harmful for reliability at scale. As it currently stands, it's basically only useful as an early warning flag to start renewing certificates _right away_ and not within a certain window, because nothing is _required_ of clients and nothing is _guaranteed_ by the server. Read the full topic to understand the full context.

Caddy uses CertMagic, which uses ACMEz, and ACMEz now has support for current-draft ARI (as Francis pointed out), but I probably won't be doing more with it in the near-future as it sounds like they'll be changing it at least a little bit (though I think more drastic changes may be required).

If nothing else, I'll use it as an early warning signal to start renewing certificates right away.

In the meantime, Caddy staples OCSP responses, and will automatically renew certificates that are revoked. Thankfully, Caddy finds this out days before the current valid OCSP staple expires, so it can renew the cert at that point while it can still serve a staple declaring the certificate is valid. It is then up to clients to honor that staple. (Revocation is broken, don't at-me. :unamused:)

Going to close this since we're waiting on processes out of our control, but will tag as deferred. Thanks!
